File: BSTFromInorderPreorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Approach: Recursive using pointers

# In this approach we are going take the advantage of the way preorder and inorder traversal is done. If we obesrver closely, we can see that out of the give lists, first element in preorder list is the root. Now the elements towards left of the index at which root (found in preorder) makes up the left subtree and the elements towards the right of the root makes up the right tree.

# So at every recursive call we will find the left and right subtree elements in inorder and preorder and pass them as an argument. So the root's left would be the tree made up of left elements in preorder and inorder and same goes for the right subtree.

# Time complexity: O(n); n = length of the given list
# Space complexity: O(n); n = length of the given list

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        self.hmap = {}
        self.idx = 0
        
        if inorder == None or len(inorder) == 0:
            return None
        
        for i in range(len(inorder)):
            self.hmap[inorder[i]] = i
            
        return self.helper(preorder, 0, len(inorder) - 1)
    
    def helper(self, preorder, start, end):
        if start > end:
            return None

        rootVal = preorder[self.idx]
        root = TreeNode(rootVal)
        self.idx += 1
        
        rootIdx = self.hmap[rootVal]
        
        root.left = self.helper(preorder, start, rootIdx - 1)
        root.right = self.helper(preorder, rootIdx + 1, end)
        return root
        
        
        

File: bst_from_inorder_preorder.cpp
class Solution {
public:
    TreeNode* buildTree(vector<int> &preorder, vector<int> &inorder) {
        if(preorder.size() == 0 || inorder.size() == 0) return NULL;
        int root_val = preorder[0] ;
        TreeNode *root = new TreeNode(root_val);
        //cout << root_val ;
        int root_index ;
        int size = inorder.size() ;
        // finding root node in inorder array
        for(int i = 0 ; i < inorder.size() ; i ++){
            if(inorder[i] == root_val){
                root_index = i ;
                break;
            }
        }
        //cout << root_index ;
        // found root index , now create sub - arrays for left and right part of the tree
        vector <int> inorder_left ;
        vector <int> inorder_right;
        vector <int> preorder_left ;
        vector <int> preorder_right;
        for(int i = 0 ; i < root_index ; i++){
            inorder_left.push_back(inorder[i]) ;
            // cout << inorder_left[i] ;
        }
        for(int j = root_index + 1 ; j < inorder.size() ; j++){
            inorder_right.push_back(inorder[j]) ;
            // cout << inorder_right[j] ;
        }
        for(int i = 1 ; i < root_index + 1  ; i++){
            preorder_left.push_back(preorder[i]) ;
            //cout << i ;
        }
        for(int i = root_index + 1 ; i < preorder.size() ; i++){
            preorder_right.push_back(preorder[i]) ;
            //cout << i ;
        }
        root -> left = buildTree(preorder_left,inorder_left) ;
        root -> right = buildTree(preorder_right,inorder_right) ;
        return root;
     }
};
